src/assets/wise5/components/draw/draw-student/draw-student.component.ts

```diff
@@ -0,0 +1,437 @@
+        }
+        this.studentDataChanged();
+      }
+    }
+  }
+
+  isUpdateImmediately(connectedComponent: any, componentState: any): boolean {
+    return (
+      connectedComponent.updateOn === 'save' ||
+      (connectedComponent.updateOn === 'submit' && componentState.isSubmit)
+    );
+  }
+
+  isComponentStateFromConnectedComponent(componentState: any): boolean {
+    return this.ProjectService.isConnectedComponent(
+      this.nodeId,
+      this.componentId,
+      componentState.componentId
+    );
+  }
+
+  isPerformUpdate(): boolean {
+    return this.isCanvasEmpty() || confirm($localize`Do you want to update the connected drawing?`);
+  }
+
+  /**
+   * Add student asset image as object in the drawing canvas
+   * @param studentAsset
+   */
+  attachStudentAsset(studentAsset: any): void {
+    fabric.Image.fromURL(studentAsset.url, (oImg: any) => {
+      oImg.scaleToWidth(this.width);
+      oImg.studentAssetId = studentAsset.id;
+      this.setBackgroundImage(studentAsset.url);
+    });
+  }
+
+  getDrawData(): any {
+    return this.drawingTool.save();
+  }
+
+  setDrawData(componentState: any): void {
+    const studentData = componentState.studentData;
+    if (studentData.submitCounter != null) {
+      this.submitCounter = studentData.submitCounter;
+    }
+    const drawData = studentData.drawData;
+    if (drawData != null && drawData != '' && drawData != '{}') {
+      this.drawingTool.load(drawData);
+    }
+  }
+
+  /**
+   * Check if the student has drawn anything
+   * @returns whether the canvas is empty
+   */
+  isCanvasEmpty(): boolean {
+    const objects = this.drawingTool.canvas.getObjects();
+    return objects == null || objects.length === 0;
+  }
+
+  addToNotebook(): void {
+    if (this.isDirty) {
+      const studentWorkSavedToServerSubscription = this.StudentDataService.studentWorkSavedToServer$.subscribe(
+        (componentState: any) => {
+          if (this.isForThisComponent(componentState)) {
+            this.addNoteWithImage(componentState.id);
+            studentWorkSavedToServerSubscription.unsubscribe();
+          }
+        }
+      );
+      this.saveButtonClicked();
+    } else {
+      const componentState = this.StudentDataService.getLatestComponentStateByNodeIdAndComponentId(
+        this.nodeId,
+        this.componentId
+      );
+      this.addNoteWithImage(componentState.id);
+    }
+  }
+
+  addNoteWithImage(componentStateId: number) {
+    const image = this.generateImageFromCanvas();
+    this.NotebookService.addNote(image, null, [componentStateId]);
+  }
+
+  generateImageFromCanvas(): any {
+    const canvasBase64Image = this.getCanvas().toDataURL('image/png');
+    return this.UtilService.getImageObjectFromBase64String(canvasBase64Image);
+  }
+
+  getCanvas(): any {
+    const canvas: any = $(`#${this.drawingToolId} canvas`);
+    if (canvas != null && canvas.length > 0) {
+      return canvas[0];
+    } else {
+      return null;
+    }
+  }
```

Comment:
This function can return null, but the place that calls it (generateImageFromCanvas()) does not null-check. Is it really possible for this function to return null? If the template exists, shouldn't it always return a canvas object?